tracer/src/Datadog.Trace/Debugger/Symbols/SymbolExtractor.cs

```diff
@@ -0,0 +1,718 @@
+                Scopes = closureScopes,
+                SourceFile = null,
+                StartLine = UnknownMethodStartLine,
+                EndLine = UnknownMethodEndLine
+            };
+
+            return methodScope;
+        }
+
+        private Model.Scope[]? GetClosureScopes(TypeDefinition typeDef, MethodDefinition methodDef)
+        {
+            Model.Scope[]? closureMethods = null;
+            try
+            {
+                var nestedTypes = typeDef.GetNestedTypes();
+                var methods = typeDef.GetMethods();
+                closureMethods = ArrayPool<Model.Scope>.Shared.Rent(methods.Count + (nestedTypes.Length * 2));
+                int index = 0;
+
+                for (int i = 0; i < nestedTypes.Length; i++)
+                {
+                    if (!DatadogMetadataReader.IsCompilerGeneratedAttributeDefinedOnType(nestedTypes[i].RowId))
+                    {
+                        continue;
+                    }
+
+                    var nestedType = MetadataReader.GetTypeDefinition(nestedTypes[i]);
+                    var generatedMethods = nestedType.GetMethods();
+                    foreach (var generatedMethodHandle in generatedMethods)
+                    {
+                        if (generatedMethodHandle.IsNil)
+                        {
+                            continue;
+                        }
+
+                        var generatedMethodDef = MetadataReader.GetMethodDefinition(generatedMethodHandle);
+                        PopulateClosureMethod(generatedMethodDef, nestedType);
+                    }
+                }
+
+                foreach (var methodHandle in methods)
+                {
+                    if (methodHandle.IsNil || methodHandle == methodDef.Handle)
+                    {
+                        continue;
+                    }
+
+                    if (!DatadogMetadataReader.IsCompilerGeneratedAttributeDefinedOnMethod(methodHandle.RowId))
+                    {
+                        continue;
+                    }
+
+                    var currentMethod = MetadataReader.GetMethodDefinition(methodHandle);
+                    PopulateClosureMethod(currentMethod, typeDef);
+                }
+
+                if (index == 0)
+                {
+                    return null;
+                }
+
+                var closureScopes = new Model.Scope[index];
+                for (int i = 0; i < index; i++)
+                {
+                    closureScopes[i] = closureMethods[i];
+                }
+
+                return closureScopes;
+
+                void PopulateClosureMethod(MethodDefinition generatedMethod, TypeDefinition ownerType)
+                {
+                    var closureMethodScope = CreateMethodScopeForGeneratedMethod(methodDef, generatedMethod, ownerType);
+                    if (closureMethodScope.HasValue)
+                    {
+                        if (index < closureMethods.Length)
```

Comment:
Isn't it true that if `index == closureMethods.Length` you'll encounter into a problem in space?
Despite that, why don't we simply enlarge the array?

Reply:
Because the initial size is big enough `methods.Count + (nestedTypes.Length * 2)`, I think it is very unlikely that we will reach this situation. That's why I preferred not to bother with enlarging the array and in the very unlikely case, write to the debug log. WDYT?

Reply:
Confused by this - if this is something we never expect to happen, shouldn't we log a WARNING if it happens, because it means we'll miss out on data?

Reply:
I don't expect it will happen but anyway there is a warning log.